src/server/frontend_wayland/wl_client.h

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright © 2021 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: William Wold <william.wold@canonical.com>
+ */
+
+#ifndef MIR_FRONTEND_WL_CLIENT_H_
+#define MIR_FRONTEND_WL_CLIENT_H_
+
+struct wl_client;
+struct wl_listener;
+struct wl_display;
+
+#include <memory>
+#include <functional>
+
+namespace mir
+{
+namespace shell
+{
+class Shell;
+}
+namespace scene
+{
+class Session;
+}
+
+namespace frontend
+{
+class SessionAuthorizer;
+
+class WlClient
+{
+public:
+    struct ConstructionCtx;
+    struct ClientCtx;
```

Comment:
Is there an advantage to declaring these structs in class scope as opposed to the anonymous namespace of the implementation file? They are not mentioned elsewhere in the header and don't need access to private parts.

Reply:
`ConstructionCtx` didn't need private access at all. `ClientCtx` was using the private constructor, but I'm not sure that was right.

Reply:
Fixed by constructing `WlClient` in `handle_client_created` (which IMO is a better place) and making that a private static function.